### PR TITLE
[NTV-201] Adding environmental commitments to new project page

### PR DESCRIPTION
**Pull request by singhhari**
# 📲 What

Adding the Environmental Commitments section to our new project page. Upon selection of the tab in the navigation selector, we will navigate to the new view controller and render a table view of environmental commitments. The very last cell (or in the case when no environmental commitments are returning from the API) we show a "footer" cell that enables users to navigate to a webview of our Environmental Resources Center.

# 🛠 How

Created a couple of new table view cells to register for our `EnvironmentalCommitmentsViewController`. Each cell and the view controller has its own view model that can micromanage its logic. Moreover we're using a delegate to enable the presentation of the `HelpWebViewController`.

# 👀 See

https://user-images.githubusercontent.com/15615702/139474040-d55fed28-300d-4ba2-a955-2a2b51bbbcd5.mp4

# ✅ Acceptance criteria

- [x] Navigate to a project in the `Popular` tab. Find one with environmental commitments like [this one](https://www.kickstarter.com/projects/flatoutgames/verdant) and verify you can see the environmental commitments and the footer cell showing a hyperlink.
- [x] Tap on the hyperlink and verify a web view controller is presented and navigates to our Environmental Resources Center.
- [ ] Similarly, find a project with no environmental commitments and verify only the footer cell is showing.

### CODE REVIEW

**Review by msadoon — CHANGES_REQUESTED** [review]
Code looks decent, there's one issue with respect to the `footer`, pretty sure it could just be a table view footer and doesn't need to be its' own cell.
Also a few other code -related comments.

Ran app on sim, found a few issues:
alignment to header title is a little off: [prototype](https://www.figma.com/file/Vk262JvQlfddgdkfXN0rx0/Native%3A-Project-Page-C5-2021?node-id=166%3A5281)


Also noticed on ipad only that after opening the web view and closing it, the contents of the entire project page are gone.


And on a similar note, I know that we restrict the FAQ page contents to if the user is logged in. Are we still doing that with environmental commitments? Seems a product question mostly, but it would be strange to show one section and not another based on log in state, unless that's the intention of the API.

File: Kickstarter-iOS/DataSources/ProjectEnvironmentalCommitmentsDataSource.swift [hunk 1]
```diff
@@ -0,0 +1,38 @@
+import KsApi
+import Library
+import UIKit
+
+internal final class ProjectEnvironmentalCommitmentsDataSource: ValueCellDataSource {
+  internal enum Section: Int {
+    case environmentalCommitments
+    case footer
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/DataSources/ProjectEnvironmentalCommitmentsDataSource.swift]: I've decided to rename this section and its respective cell from `Footer` to `Disclaimer`. I agree that using the word footer is misleading but I do not think it is worth the effort of changing this implementation given the significant re-factor I'm going to have to work on right after this.

File: Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+      string: "to learn how Kickstarter encourages sustainable practices.",
+      attributes: regularFontAttribute
+    )
+
+    guard let environmentLink = HelpType.environment
+      .url(withBaseUrl: AppEnvironment.current.apiService.serverConfig.webBaseUrl)?.absoluteString else {
+      return learnMoreString
+    }
+
+    let environmentString = "<a href=\(environmentLink)>Visit our Environmental Resources Center</a>"
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift]: I'm unsure what the suggestion is here? There already is a TODO at the top of this method that suggests the string needs to be internationalized, I'll add another line of details to it?  

More importantly though, the string we submit will very likely be these two, word-for-word. The one with the anchor tag will need to be submitted exactly like that so the translation is wrapped by that tag and the UITextView delegate method can detect it. If you look at the `attributedDisclaimerText` method, the String being localized is using this same method and the argument to the method is a URL that gets embedded in the anchor tag. We will be employing the exact same method here. I think once we get the translations back we can clean up the method but for the most part I don't think there is a way to simplify applying separate attributes to both strings and then combining them. Open to suggestions.

File: Kickstarter-iOS/Views/Controllers/ProjectEnvironmentalCommitmentsViewControllerTests.swift [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        id: 0
+      ),
+      ProjectEnvironmentalCommitment(description: "hello world", category: .longLastingDesign, id: 1),
+      ProjectEnvironmentalCommitment(
+        description: "Lorem ipsum",
+        category: .reusabilityAndRecyclability,
+        id: 2
+      ),
+      ProjectEnvironmentalCommitment(description: "blah blah blah", category: .sustainableDistribution, id: 3)
+    ]
```
[the remark this answers is not in the document]
Comment (singhhari) [reply Kickstarter-iOS/Views/Controllers/ProjectEnvironmentalCommitmentsViewControllerTests.swift]: Good catch. Updated.

**Comment by singhhari** [comment]
> Code looks decent, there's one issue with respect to the `footer`, pretty sure it could just be a table view footer and doesn't need to be its' own cell. Also a few other code -related comments.
> 
> Ran app on sim, found a few issues: alignment to header title is a little off: [prototype](https://www.figma.com/file/Vk262JvQlfddgdkfXN0rx0/Native%3A-Project-Page-C5-2021?node-id=166%3A5281) 
> 
> Also noticed on ipad only that after opening the web view and closing it, the contents of the entire project page are gone. 
> 
> And on a similar note, I know that we restrict the FAQ page contents to if the user is logged in. Are we still doing that with environmental commitments? Seems a product question mostly, but it would be strange to show one section and not another based on log in state, unless that's the intention of the API. 

Nice eye here. I was adding a small margin without realizing it was slightly off. Good catch.

As for the log in state issue, I'm not sure why we have a difference here. That's definitely a product question but my assumption is that the Environmental Resources center is available to unauthenticated users anyway so there's no reason to restrict access to it. Asking a question to the creator does seem like something that requires the context of a logged in user so the creator knows who is asking.

File: Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (msadoon) [reply Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift]: Yea its' fine for now, my suggestion was to add the `//TODO:` here to address that we're not driving this `String` through `Localizable.strings`, as done with `atttributedDisclaimerText` right? So yes we need to wait for translations for that which is fine, I just want to be aware that we don't won't two ways to generate links for a string, `Localizable` _and_ `a href=` scattered elsewhere, just a convention really, but helps keep our code base clean. Also I don't see examples of `a href` anywhere else in the code, just trying to be careful about introducing new methods of generating `Strings` is all :)

File: Library/ViewModels/ProjectEnvironmentalCommitmentsViewModelTests.swift [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+        id: 0
+      ),
+      ProjectEnvironmentalCommitment(description: "hello world", category: .longLastingDesign, id: 1),
+      ProjectEnvironmentalCommitment(
+        description: "Lorem ipsum",
+        category: .reusabilityAndRecyclability,
+        id: 2
+      ),
+      ProjectEnvironmentalCommitment(description: "blah blah blah", category: .sustainableDistribution, id: 3)
+    ]
```
[the remark this answers is not in the document]
Comment (msadoon) [reply Library/ViewModels/ProjectEnvironmentalCommitmentsViewModelTests.swift:90]: I think you may have forgotten to push the change here.

**Review by msadoon — APPROVED** [review]
Ok looks good,

Just push that last change with the tests property being declared globally and we're good I think. We can address the Strings issue once we do the actual Strings upload.

Great job!

File: Library/ViewModels/ProjectEnvironmentalCommitmentsViewModelTests.swift (hunk 4 above)
[the remark this answers is not in the document]
Comment (singhhari) [reply Library/ViewModels/ProjectEnvironmentalCommitmentsViewModelTests.swift:90]: Yup lol sry about that. Pushing it up in this last commit.

File: Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (msadoon) [reply Kickstarter-iOS/Views/Cells/ProjectEnvironmentalCommitmentFooterCell.swift]: Cool, just wanted to point out while it ofcourse used in multiple places in our codebase, the only place its' used outside of `Strings` or `Localizable` is `DisclaimerCell` :)

**Pull request merged**